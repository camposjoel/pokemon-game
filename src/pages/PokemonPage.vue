<template>
  <h1 v-if="!pokemon">Espere por favor</h1>
  <div v-else>
    <h1>¿Quién es este pokemon?</h1>
    
    <PokemonPicture
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon"
    />
    <PokemonOptions
      :pokemons="pokemonList"
      @selection="checkAnswer"  
    />

    <template v-if="showAnswer">
      <div class="newGameContainer">
        <h2 class="fade-in">{{message}}</h2>
        <button @click="newGame" class="newGameButton">
          Nuevo Juego
        </button>
      </div>
    </template>
  </div>
</template>

<script>
import PokemonOptions from "../components/PokemonOptions.vue"
import PokemonPicture from "../components/PokemonPicture.vue"
import getPokemonsOptions from "../helpers/getPokemonOptions"


export default {
  components: {
    PokemonOptions,
    PokemonPicture
  },
  data() {
    return {
      pokemonList: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonList() {
      this.pokemonList = await getPokemonsOptions()

      const rndInt = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonList[rndInt]
    },
    checkAnswer(selectedPokemonId) {
      this.showPokemon = true
      this.showAnswer = true

      if (selectedPokemonId === this.pokemon.id) {
        this.message = `Correcto, ${this.pokemon.name}`
      } else {
        this.message = `Oops, era ${this.pokemon.name}`
      }
    },
    newGame() {
      this.showPokemon = false
      this.showAnswer = false
      this.pokemon = null
      this.mixPokemonList()
    }
  },
  mounted() {
    this.mixPokemonList()
  }
}
</script>

<style scoped>
h1 {
  font-size: 3rem;
}

.newGameContainer {
  position: absolute;
  bottom: 25px;
  left: 0;
  right: 0;
  margin-left: auto;
  margin-right: auto;
}

.newGameButton {
  border: none;
  height: 40px;
  width: 140px;
  background-color: lightgreen;
  color: #2c3e50;
  font-family: 'Josefin Sans';
  border-radius: 8px;
  font-size: 16px;
  font-weight: bold;
  cursor: pointer;
}

.newGameButton:hover {
  box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;
}

</style>
