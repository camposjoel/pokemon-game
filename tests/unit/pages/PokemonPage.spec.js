// @vitest-environment happy-dom
import { describe, expect, test, beforeEach, spyOn } from "vitest"
import { mount, shallowMount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {
  let wrapper
  
  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })
  
  test('Should match with snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Should call mixPokemonList in mount', () => {
    const mixPokemonArraySpy = spyOn(PokemonPage.methods, 'mixPokemonList')
    shallowMount(PokemonPage)
    expect(mixPokemonArraySpy).toHaveBeenCalled()
  })

  test('Should match with snapshot when pokemons are loaded', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonList: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Should show the PokemonPicture and PokemonOptions components', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonList: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    const pictureWrapper = wrapper.find('pokemon-picture-stub')
    const optionsWrapper = wrapper.find('pokemon-options-stub')
    expect(pictureWrapper.exists()).toBeTruthy()
    expect(optionsWrapper.exists()).toBeTruthy()
    expect(pictureWrapper.attributes('pokemonid')).toBe('1')
    expect(optionsWrapper.attributes('pokemons')).toBeTruthy()
  })

  test('CheckAnswer should work', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonList: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })

    await wrapper.vm.checkAnswer(1)
    expect(wrapper.find('h2').exists()).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.find('h2').text()).toBe(`Correcto, ${pokemons[0].name}`)

    await wrapper.vm.checkAnswer(2)
    expect(wrapper.find('h2').text()).toBe(`Oops, era ${pokemons[0].name}`)
  })
  
})