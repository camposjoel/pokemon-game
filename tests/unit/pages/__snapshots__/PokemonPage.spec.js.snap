// Vitest Snapshot v1

exports[`PokemonPage Component > Should match with snapshot 1`] = `"<h1 data-v-f44bbbe4=\\"\\">Espere por favor</h1>"`;

exports[`PokemonPage Component > Should match with snapshot when pokemons are loaded 1`] = `
"<div data-v-f44bbbe4=\\"\\">
  <h1 data-v-f44bbbe4=\\"\\">¿Quién es este pokemon?</h1>
  <pokemon-picture-stub pokemonid=\\"1\\" showpokemon=\\"false\\" data-v-f44bbbe4=\\"\\"></pokemon-picture-stub>
  <pokemon-options-stub pokemons=\\"[object Object],[object Object],[object Object],[object Object]\\" data-v-f44bbbe4=\\"\\"></pokemon-options-stub>
  <!--v-if-->
</div>"
`;
